In [ ]:
survey <- read.csv("C:/Users/tyler/OneDrive/Documents/GitHub/work_to_show/Classes Spring 2026/BDA Spring 26/Datasets/survey.csv")
head(survey)
colnames(survey)

In [ ]:
# install.packages("arules") #run only once
library("arules")

#identify transactions
transactions <- as(split(survey$Answer, survey$Stdnt_Nbr),"transactions")

# use the apriori algorithm for association analysis
# specify the minimum support and minimum confidence
# maxlen=2 tells R to only consider pairwise associations
rules<-apriori(transactions,parameter=list(support=0.02,confidence=0.1,maxlen=2))

#include the odds ratio, use the Haldane-Anscombe correction, sort by odds ratio
ci <- confint(rules, "oddsRatio", smoothCounts = .5)

quality(rules) <- cbind(
quality(rules),
oddsRatio = interestMeasure(rules, "oddsRatio", smoothCounts = .5),
oddsRatio = ci)

rules <- sort(rules, by = "oddsRatio")
inspect(rules)

In [ ]:
music <- read.csv("C:/Users/tyler/OneDrive/Documents/GitHub/work_to_show/Classes Spring 2026/BDA Spring 26/Datasets/music.csv")
colnames(music)
head(music)

In [ ]:
# Customers who added “TheShins” to their favorites list were 3.8 times more likely to add “VampireWeekend” to their favorites list.
#install.packages("arules") #run only once
library("arules")
#identify transactions
# transactions <- as(split(df$y, df$x),"transactions")
transactions <- as(split(music$artist, music$CustomerID),"transactions")
#use the apriori algorithm for association analysis
#specify the minimum support and minimum confidence
#maxlen=2 tells R to only consider pairwise associations
rules<-apriori(transactions,parameter=list(support=0.02,confidence=0.1,maxlen=2))
#include the odds ratio, use the Haldane-Anscombe correction, sort by odds ratio
ci <- confint(rules, "oddsRatio", smoothCounts = .5)
quality(rules) <- cbind(
quality(rules),
oddsRatio = interestMeasure(rules, "oddsRatio", smoothCounts = .5),
oddsRatio = ci)
rules <- sort(rules, by = "oddsRatio")
inspect(rules)

# homework 

In [5]:
df <- read.csv("C:/Users/tyler/OneDrive/Documents/GitHub/work_to_show/Classes Spring 2026/BDA Spring 26/Datasets/mb.csv")
colnames(df)
head(df)

[1] "visitid"  "itemtype"

,visitid,itemtype
,<int>,<chr>
1,100001,Antacid
2,100001,Bread
3,100001,Candy
4,100001,CatFood
5,100001,Chips
6,100001,Cookies


In [6]:
#install.packages("arules") #run only once
library("arules")
#identify transactions
# transactions <- as(split(df$y, df$x),"transactions")
transactions <- as(split(df$itemtype, df$visitid),"transactions")

#use the apriori algorithm for association analysis
#specify the minimum support and minimum confidence
#maxlen=2 tells R to only consider pairwise associations
rules<-apriori(transactions,
               parameter=
                 list(support=0.02,confidence=0.1,maxlen=2)
               )

#include the odds ratio, use the Haldane-Anscombe correction, sort by odds ratio
ci <- confint(rules, "oddsRatio", smoothCounts = .5)
quality(rules) <- cbind(
  quality(rules),
  oddsRatio = interestMeasure(rules, "oddsRatio", smoothCounts = .5),
  oddsRatio = ci)

rules <- sort(rules, by = "oddsRatio")
inspect(rules)

# Convert rules to a data frame and export to CSV
rules_df <- as(rules, "data.frame")

# Also add LHS and RHS as separate columns for readability
rules_df$lhs <- labels(lhs(rules))
rules_df$rhs <- labels(rhs(rules))

write.csv(rules_df, "C:/Users/tyler/OneDrive/Documents/GitHub/work_to_show/Classes Spring 2026/BDA Spring 26/mb_rules.csv", row.names = FALSE)

cat("Exported", nrow(rules_df), "rules to CSV.\n")

Apriori

Parameter specification:
 confidence minval smax arem  aval originalSupport maxtime support minlen
        0.1    0.1    1 none FALSE            TRUE       5    0.02      1
 maxlen target  ext
      2  rules TRUE

Algorithmic control:
 filter tree heap memopt load sort verbose
    0.1 TRUE TRUE  FALSE TRUE    2    TRUE

Absolute minimum support count: 3 

set item appearances ...[0 item(s)] done [0.00s].
set transactions ...[20 item(s), 188 transaction(s)] done [0.00s].
sorting and recoding items ... [18 item(s)] done [0.00s].
creating transaction tree ... done [0.00s].
checking subsets of size 1 2

Warning message in apriori(transactions, parameter = list(support = 0.02, confidence = 0.1, :
"Mining stopped (maxlen reached). Only patterns up to a length of 2 returned!"


 done [0.00s].
writing ... [269 rule(s)] done [0.00s].
creating S4 object  ... done [0.00s].
      lhs                   rhs                support    confidence coverage  
[1]   {Ammunition}       => {BeefJerky}        0.02659574 0.5555556  0.04787234
[2]   {BeefJerky}        => {Ammunition}       0.02659574 0.2380952  0.11170213
[3]   {ToothPaste}       => {Cookies}          0.03723404 1.0000000  0.03723404
[4]   {BeefJerky}        => {Chips}            0.10638298 0.9523810  0.11170213
[5]   {Chips}            => {BeefJerky}        0.10638298 0.1612903  0.65957447
[6]   {InkPen}           => {Milk}             0.07446809 0.9333333  0.07978723
[7]   {Milk}             => {InkPen}           0.07446809 0.1238938  0.60106383
[8]   {InkPen}           => {Chips}            0.07446809 0.9333333  0.07978723
[9]   {Chips}            => {InkPen}           0.07446809 0.1129032  0.65957447
[10]  {ToothPaste}       => {Antacid}          0.02127660 0.5714286  0.03723404
[11]  {ToothPaste}       =>

# two most associated items: 
Ammunition and Beef Jerky
Odds Ratio: 12.11
Support 2.66%
Confidence 55.6%

Customers who purchase ammunition are approximately 12.11 times more likely to purchase beef jerky commpared to customers who don't purchase ammo. 

Business Application: Change the store Layout, kind of make them close together. Do a bundle promotion or something like that. 

Does not support toothpaste and cookies together. Odds ratio is 9.93 support is 3.7% and was 7 transactions. 